### DISCUSSION TITLE
Backup: Check data directory before backing up

### DISCUSSION THREAD

**Discussion by gordon-zhao** [nextcloud/all-in-one#4006] [Ideas] [answered]
### Is your feature request related to a problem? Please describe. 

For those who has data directory resides in different drives/locations, there is a risk of having an archive with empty data directory when (not realizing) the data directory is unmounted at the time of backup. Restoring such archive with data directory mounted will nuke existing data directory without notice. I learned this in the hard way (yes, that's how I lost my data but lucky able to recover from journals)

### Describe the solution you'd like

1. Having a mechanism for checking if current data directory is non-empty and valid before backing up. 
2. Make a copy of existing data directory before attempting to restore an archive.

**Comment by szaimen** [comment, marked as answer]
Hi, thanks for your ideas!

> * Having a mechanism for checking if current data directory is non-empty and valid before backing up.

We already have such a mechanism here: https://github.com/nextcloud/all-in-one/blob/2197a3bdbf70a81e51fc4a385b445e144156f35a/Containers/borgbackup/backupscript.sh#L75-L79
So apparently this did not work for you. Most likely the drive was partially still mounted when the backup archive was created which would lead to some files/folders still being discovered in this directory which is something that would need to be fixed outside of AIO as it sounds like a kernel bug to me.

> * Make a copy of existing data directory before attempting to restore an archive.

This is unfortunately obviously out of scope since it would double the needed storage space.

So currently I don't see how we can further improve this in AIO.

**Reply by gordon-zhao** [reply]
Thanks for the comment! My further investigation confirmed that my data mountpoint is non-empty and somehow mount.cifs allows overmounting on non-empty directory (weird...). 

However, the mountpoint was supposed to be empty as this was how I set it up, and nextcloud shouldn't write any data to empty directory as .ocdata is not present. But, I found 1 folder and 2 files in the mountpoint: files_external, fingerprint.update and skip.update. 

This is probably what happened: 
1. At some time in the past, I updated the aio without data folder being mounted. 
2. The update procedure left some files in the empty data folder (the mountpoint) and nextcloud complains for not having Appdata presented. 
3. Then I remounted the data folder without realizing the overmounting issue. 
4. This time, I logged in to AIO and attempted to restart all container as nextcloud was not starting (not realizing the data was not mounted due to -nofail option), but AIO would not start container without updating and prompted for creating a backup before the update. 

That's probably how I managed to create an archive with *empty* data folder. 

If I understood correctly, the current mechanism does not explicitly check the datadir as nextcloud would do (looking for .ocdata, Appdata, etc.). Is it a good idea to implement an explicit check on datadir if NEXTCLOUD_DATADIR is defined? For example:
```
if ! [ -f "/nextcloud_aio_volumes/nextcloud_aio_nextcloud_data/.ocdata" ]; then
    echo "/nextcloud_aio_volumes/nextcloud_aio_nextcloud_data is not a valid datadir which should not happen!"
    exit 1
fi
```

**Reply by szaimen** [reply]
See https://github.com/nextcloud/all-in-one/pull/4021

**Reply by szaimen** [reply]
This is now released with v7.10.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by gordon-zhao** [reply]
Awesome. Thank you for accepting the suggestion!
